### PR TITLE
eth/tracers/logger: use omitempty to reduce log bloat 

### LINKED ISSUE #24487: eth/tracers/logger: use omitempty in error JSON tag to avoid unnecessary log bloat
#### System information

Geth version: `geth version` Not Applicable
OS & Version: Windows/Linux/OSX Not Applicable
Commit hash : 687e4dc8555b3b53fd8d4a4974b8f8f4b594b552

#### Expected behaviour
Lack of error strings should NOT log `"error":""` which bloats logs and can actually be confusing when you grep for "error" in logs when things are failing.

#### Actual behaviour
Looking at such logs like
```shell
{"pc":4716,"op":80,"gas":"0x688e5","gasCost":"0x2","memory":"0x","memSize":352,"stack":["0xfa31de01","0x29a","0x7d0","0xabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcd","0x80","0x534","0xc0","0x5f646b5592c4159d4963fbf901f81144ef53cc51a7082435e3a6120a45002ff7","0x15dc","0x34","0x5f646b5592c4159d4963fbf901f81144ef53cc51a7082435e3a6120a45002ff7"],"returnData":"0x","depth":2,"refund":0,"opName":"POP","error":""}
{"pc":4717,"op":80,"gas":"0x688e3","gasCost":"0x2","memory":"0x","memSize":352,"stack":["0xfa31de01","0x29a","0x7d0","0xabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcd","0x80","0x534","0xc0","0x5f646b5592c4159d4963fbf901f81144ef53cc51a7082435e3a6120a45002ff7","0x15dc","0x34"],"returnData":"0x","depth":2,"refund":0,"opName":"POP","error":""}
{"pc":4718,"op":86,"gas":"0x688e1","gasCost":"0x8","memory":"0x","memSize":352,"stack":["0xfa31de01","0x29a","0x7d0","0xabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcd","0x80","0x534","0xc0","0x5f646b5592c4159d4963fbf901f81144ef53cc51a7082435e3a6120a45002ff7","0x15dc"],"returnData":"0x","depth":2,"refund":0,"opName":"JUMP","error":""}
{"pc":5596,"op":91,"gas":"0x688d9","gasCost":"0x1","memory":"0x","memSize":352,"stack":["0xfa31de01","0x29a","0x7d0","0xabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcd","0x80","0x534","0xc0","0x5f646b5592c4159d4963fbf901f81144ef53cc51a7082435e3a6120a45002ff7"],"returnData":"0x","depth":2,"refund":0,"opName":"JUMPDEST","error":""}
{"pc":5597,"op":97,"gas":"0x688d8","gasCost":"0x3","memory":"0x","memSize":352,"stack":["0xfa31de01","0x29a","0x7d0","0xabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcd","0x80","0x534","0xc0","0x5f646b5592c4159d4963fbf901f81144ef53cc51a7082435e3a6120a45002ff7"],"returnData":"0x","depth":2,"refund":0,"opName":"PUSH2","error":""}
{"pc":5600,"op":97,"gas":"0x688d5","gasCost":"0x3","memory":"0x","memSize":352,"stack":["0xfa31de01","0x29a","0x7d0","0xabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcd","0x80","0x534","0xc0","0x5f646b5592c4159d4963fbf901f81144ef53cc51a7082435e3a6120a45002ff7","0x15ef"],"returnData":"0x","depth":2,"refund":0,"opName":"PUSH2","error":""}
{"pc":5603,"op":97,"gas":"0x688d2","gasCost":"0x3","memory":"0x","memSize":352,"stack":["0xfa31de01","0x29a","0x7d0","0xabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcd","0x80","0x534","0xc0","0x5f646b5592c4159d4963fbf901f81144ef53cc51a7082435e3a6120a45002ff7","0x15ef","0x15e7"],"returnData":"0x","depth":2,"refund":0,"opName":"PUSH2","error":""}
```

Notice that each log line has `"error":""` per
Screen Shot 2022-03-01 at 12 31 39 PM

#### Steps to reproduce the behaviour
Just run the tracers

This bug was sponsored by Tharsis / evmos.org

### PR DESCRIPTION
**Pull request by holiman**
This PR replaces #24491. I pushed some fixes on top, but do not have permissions to push to the submitter's repo. Also rebased on master. 

Closes #24487

### CODE REVIEW

File: eth/tracers/logger/logger.go [hunk 1]
```diff
@@ -80,10 +80,10 @@ type StructLog struct {
 type structLogMarshaling struct {
 	Gas         math.HexOrDecimal64
 	GasCost     math.HexOrDecimal64
-	Memory      hexutil.Bytes
-	ReturnData  hexutil.Bytes
-	OpName      string `json:"opName"` // adds call to OpName() in MarshalJSON
-	ErrorString string `json:"error"`  // adds call to ErrorString() in MarshalJSON
+	Memory      hexutil.Bytes `json:",omitempty"`
+	ReturnData  hexutil.Bytes `json:",omitempty"`
+	OpName      string        `json:"opName"`          // adds call to OpName() in MarshalJSON
```
Comment (s1na) [inline eth/tracers/logger/logger.go:85]: Is OpName empty in any case?

File: eth/tracers/logger/logger.go [hunk 2]
```diff
@@ -80,10 +80,10 @@ type StructLog struct {
 type structLogMarshaling struct {
 	Gas         math.HexOrDecimal64
 	GasCost     math.HexOrDecimal64
-	Memory      hexutil.Bytes
-	ReturnData  hexutil.Bytes
-	OpName      string `json:"opName"` // adds call to OpName() in MarshalJSON
-	ErrorString string `json:"error"`  // adds call to ErrorString() in MarshalJSON
+	Memory      hexutil.Bytes `json:",omitempty"`
```
Comment (s1na) [inline eth/tracers/logger/logger.go:83]:
```
❯ ./evm --debug --json --code "0x60006000" run                                                                                                                                 <<<
{"pc":0,"op":96,"gas":"0x2540be400","gasCost":"0x3","memory":"0x","memSize":0,"stack":[],"returnData":"0x","depth":1,"refund":0,"opName":"PUSH1"}
{"pc":2,"op":96,"gas":"0x2540be3fd","gasCost":"0x3","memory":"0x","memSize":0,"stack":["0x0"],"returnData":"0x","depth":1,"refund":0,"opName":"PUSH1"}
{"pc":4,"op":0,"gas":"0x2540be3fa","gasCost":"0x0","memory":"0x","memSize":0,"stack":["0x0","0x0"],"returnData":"0x","depth":1,"refund":0,"opName":"STOP"}
```

`memory` and `returnData` still have `0x` when they're empty so they're not really omitted

Comment (holiman) [reply eth/tracers/logger/logger.go:83]: Here's how I want it to work (not sure it does): 
- If you run it with memory enabled, and returndata enabled, then you get `"memory":"0x"` if memory is indeed blank. 
- If you run it with memory disabled, so then we don't even output `"memory"` at all.

Comment (holiman) [reply eth/tracers/logger/logger.go:83]: >  (not sure it does):

**Narrator: And as it doesn't**

Comment (s1na) [reply eth/tracers/logger/logger.go:83]: The type `Bytes` seems to encode `0x` even if the slice is nil

Comment (holiman) [reply eth/tracers/logger/logger.go]: > The type Bytes seems to encode 0x even if the slice is nil

Yes.. `omitempty` might just look at the encoded return value, and check if it's empty. We _don't_ want to change the marshaller for `types.Bytes`, so I think we'll have to just pick one behaviour.

**Review by s1na — APPROVED** [review]
LGTM

File: eth/tracers/logger/logger_test.go [hunk 3]
```diff
@@ -17,9 +17,11 @@
 package logger
 
 import (
+	"encoding/json"
 	"math/big"
 	"testing"
 
+	"fmt"
```
Comment (rjl493456442) [inline eth/tracers/logger/logger_test.go:24]: Please fix the import

**Review by rjl493456442 — APPROVED** [review]
Please fix one import issue, otherwise lgtm

**Pull request merged**